invoker/core/src/main/java/com/google/cloud/functions/invoker/BackgroundFunctionExecutor.java

```diff
@@ -76,23 +121,31 @@ public static BackgroundFunctionExecutor forClass(Class<?> functionClass) {
           "Could not construct an instance of " + functionClass.getName() + ": " + e, e);
     }
     FunctionExecutor<?> executor;
-    if (instance instanceof RawBackgroundFunction) {
-      executor = new RawFunctionExecutor((RawBackgroundFunction) instance);
-    } else {
-      BackgroundFunction<?> backgroundFunction = (BackgroundFunction<?>) instance;
-      @SuppressWarnings("unchecked")
-      Class<? extends BackgroundFunction<?>> c =
-          (Class<? extends BackgroundFunction<?>>) backgroundFunction.getClass();
-      Optional<Type> maybeTargetType = backgroundFunctionTypeArgument(c);
-      if (!maybeTargetType.isPresent()) {
-        // This is probably because the user implemented just BackgroundFunction rather than
-        // BackgroundFunction<T>.
-        throw new RuntimeException(
-            "Could not determine the payload type for BackgroundFunction of type "
-                + instance.getClass().getName()
-                + "; must implement BackgroundFunction<T> for some T");
-      }
-      executor = new TypedFunctionExecutor<>(maybeTargetType.get(), backgroundFunction);
+    switch (functionKind) {
+      case RAW_BACKGROUND:
+        executor = new RawFunctionExecutor((RawBackgroundFunction) instance);
+        break;
+      case BACKGROUND:
+        BackgroundFunction<?> backgroundFunction = (BackgroundFunction<?>) instance;
+        @SuppressWarnings("unchecked")
+        Class<? extends BackgroundFunction<?>> c =
+            (Class<? extends BackgroundFunction<?>>) backgroundFunction.getClass();
+        Optional<Type> maybeTargetType = backgroundFunctionTypeArgument(c);
+        if (!maybeTargetType.isPresent()) {
+          // This is probably because the user implemented just BackgroundFunction rather than
+          // BackgroundFunction<T>.
+          throw new RuntimeException(
+              "Could not determine the payload type for BackgroundFunction of type "
+                  + instance.getClass().getName()
+                  + "; must implement BackgroundFunction<T> for some T");
+        }
+        executor = new TypedFunctionExecutor<>(maybeTargetType.get(), backgroundFunction);
+        break;
+      case CLOUD_EVENTS:
+        executor = new CloudEventFunctionExecutor((ExperimentalCloudEventsFunction) instance);
+        break;
```

Comment:
I don't know how this repo is structured, but typically the CloudEvent support/signature is completely tangential from Event/Background function support and should be in it's own file.

Reply:
Well, we're going to be adding logic to translate from one to the other, so that for example a legacy BackgroundFunction can be triggered by an incoming CloudEvent, so ultimately there's going to be some mixing.